payments/payments.go

```diff
@@ -0,0 +1,41 @@
+package payments
+
+import (
+	"context"
+	"net/http"
+)
+
+const (
+	StripeProvider = "stripe"
+	PayPalProvider = "paypal"
+)
+
+// Provider represents a payment provider that can optionally charge, refund,
+// preauthorize payments.
+type Provider interface {
```

Comment:
Ok how about we could try having methods on the provider that take all the expected params. Then the closures and builder methods just extract those. 

For instance, in the paypal charger instance we could do it like this: 
```
func (pp *paypalProvider) NewCharger(...)(Charger, error) {
  // extract from the context like you're doing 
  params := ....
  return func(amount int, currency string) ... {
    return pp.charge(params, amount, currency)
  }
}
```